discovery/bootstrapper.go

```diff
@@ -124,102 +124,130 @@ func (c *ChannelGraphBootstrapper) SampleNodeAddrs(numAddrs uint32,
-	// nodes, or we reach our limit. We ensure that we meet the randomly
-	// sample constraint as we maintain an xor accumulator to ensure we
-	// randomly sample nodes independent of the iteration of the channel
-	// graph.
-	sampleAddrs := func() ([]*lnwire.NetAddress, error) {
-		var (
-			a []*lnwire.NetAddress
-
-			// We'll create a special error so we can return early
-			// and abort the transaction once we find a match.
-			errFound = fmt.Errorf("found node")
-		)
+	// graph and assign each one a lottery number as a sequence of bytes.
+	// To maintain the random sampling criteria, the lottery number for each
+	// node is calculated as the hash of the node's public key and our
+	// accumulator. The lowest `numAddrs` lottery numbers are kept as winners.
+	var (
+		// candidate bootstrap nodes
+		nodes []autopilot.Node
+		// corresponding lottery numbers for each node above, maintained in
+		// ascending sorted order
+		hashes [][]byte
+	)
+	err := c.chanGraph.ForEachNode(func(node autopilot.Node) error {
+		nID := autopilot.NewNodeID(node.PubKey())
 
-		err := c.chanGraph.ForEachNode(func(node autopilot.Node) error {
-			nID := autopilot.NewNodeID(node.PubKey())
-			if _, ok := c.tried[nID]; ok {
-				return nil
-			}
+		// pass by nodes which we should ignore
+		if _, ok := c.tried[nID]; ok {
+			return nil
+		}
 
-			// We'll select the first node we come across who's
-			// public key is less than our current accumulator
-			// value. When comparing, we skip the first byte as
-			// it's 50/50. If it isn't less, than then we'll
-			// continue forward.
-			nodePub := node.PubKey().SerializeCompressed()[1:]
-			if bytes.Compare(c.hashAccumulator[:], nodePub) > 0 {
-				return nil
+		// pass by nodes with no TCP addresses, since we only bootstrap with TCP
+		hasTCPAddr := false
+		for _, nodeAddr := range node.Addrs() {
+			if _, ok := nodeAddr.(*net.TCPAddr); ok {
+				hasTCPAddr = true
+				break
 			}
+		}
+		if !hasTCPAddr {
+			return nil
+		}
 
-			for _, nodeAddr := range node.Addrs() {
-				// If we haven't yet reached our limit, then
-				// we'll copy over the details of this node
-				// into the set of addresses to be returned.
-				tcpAddr, ok := nodeAddr.(*net.TCPAddr)
-				if !ok {
-					// If this isn't a valid TCP address,
-					// then we'll ignore it as currently
-					// we'll only attempt to connect out to
-					// TCP peers.
-					return nil
-				}
-
-				// At this point, we've found an eligible node,
-				// so we'll return early with our shibboleth
-				// error.
-				a = append(a, &lnwire.NetAddress{
-					IdentityKey: node.PubKey(),
-					Address:     tcpAddr,
-				})
+		// calculate a hash of this node's pubkey and our accumulator, which
+		// acts as its lottery number into our sampling
+		h := sha256.Sum256(
+			append(c.hashAccumulator[:], node.PubKey().SerializeCompressed()...),
+		)
+		// convert array to a slice
+		hash := h[:]
+
+		// determine whether this node should be included
+		if len(nodes) == 0 {
```

Comment:
If I understand the solution correctly, we'll assign each node a "lottery number" and pick the lowest `numAddrs`candidates. Rather than maintaining the list of "lottery numbers" in order as we go, we could just compute the hash for each node and add it to the list, sort the list in descending/ascending order after we're done, and finally pick the first/last `numAddrs` of the list.